### ISSUE TITLE
Provide `original` vocabulary to presave and update vocabulary hooks

### ISSUE THREAD

**Issue by argiepiano** [backdrop/backdrop-issues#5475]
## Description of the need

When vocabularies were entities in D7, whenever you updated a vocabulary, the presave and update hooks received the original (unchanged) vocabulary under $entity->original. In Backdrop, when vocabularies were demoted to config classes, this feature was removed.

It would be very helpful to bring back this feature. It's not hard to do. 

## Proposed solution

In the `save` method of TaxonomyVocabulary, we'd need to load the original vocabulary before it's saved, and add it to the vocabulary object under `original`. We would also need to create a `vocabulary_load_unchanged()` function to circumvent the cache and load the original vocabulary before it's saved.

I can probably provide a PR for this soon. But I'd like to ask if there is interest in this feature?

**Comment by argiepiano** [comment]
PR https://github.com/backdrop/backdrop/pull/3920

Ended up doing this in a slightly different way. It's such a simple addition and can really be helpful (for example, for Rules, providing continuity with D7 usage).

**Comment by indigoxela** [comment]
> It would be very helpful to bring back this feature...  (for example, for Rules, providing continuity with D7 usage).

Agreed, although I don't have an immediate need for it.

I tested locally with a mini-module that implements hook_taxonomy_vocabulary_presave - works as expected. :+1: Let's see if we can get a little more feedback here.

**Comment by quicksketch** [comment]
PR in concept looks good. Other entities like user, node, and comment have `original` when saving updates, and some functionality assumes that all entities have this key, including `file_field_update()` that tracks file usage and tokens in `entity_tokens()`. I think it would be a good idea to formalize this assumption and just say it should exist for all entities.

One thing that bothers me is that `original` is not defined as a property anywhere, so assigning it a value like this throws a warning in IDEs like PHPStorm. I think it would be suitable to define it in the `abstract class Entity`, and give it a docblock like this:

```
  /**
   * Contains the Entity object prior to changes during update operations.
   *
   * Outside of updates, this property is usually NULL.
   *
   * @var Entity
   */
  public $original
```

This does have the downside of always having `$entity->original` existing as a property, even though it will be empty most of the time. But it has the upside of helping auto-complete and giving that property documentation.

**Comment by argiepiano** [comment]
> One thing that bothers me is that original is not defined as a property anywhere, so assigning it a value like this throws a warning in IDEs like PHPStorm. I think it would be suitable to define it in the abstract class Entity, and give it a docblock 

@quicksketch, that makes sense. Should we also add a similar property and a docblock to class `TaxonomyVocabulary`?

I can add those to the PR.

**Comment by argiepiano** [comment]
I have added the property $original to both `Entity` and `TaxonomyVocabulary`.

**Comment by quicksketch** [comment]
This looks good to me. Seems this might be considered a bug. Should it go into the next bug fix release 1.21.2?

Whether we do it as a follow-up or here, I think it might be beneficial to put the same `public $original` docblock in all the other entity classes as well.

**Comment by argiepiano** [comment]
> Whether we do it as a follow-up or here, I think it might be beneficial to put the same public $original docblock in all the other entity classes as well.

Do you mean add it in Node, TaxonomyTerm, User,  etc?  (TaxonomyVocabulary is technically not an entity class, since vocabularies are not entities).

**Comment by quicksketch** [comment]
Yeah, then each would get the same benefit of autocomplete on their
respective entity type.

**Comment by argiepiano** [comment]
> Yeah, then each would get the same benefit of autocomplete on their
respective entity type.

But isn't enough (for autofilling) that the property `$original` is defined in their parent class (Entity)? At least in my IDE it is - the autocomplete happens correctly in Node, etc. Is there a need to override it/redefine it in each child as you are suggesting?

**Comment by quicksketch** [comment]
If we add the docblock with a different `@var TaxonomyVocabulary` line, then each entity type will autocomplete on the more specific properties of their respective entities. So a `@var Node` for the `$node->original` would make it so that node-only properties and methods would auto complete.

**Comment by klonos** [comment]
> Seems this might be considered a bug. Should it go into the next bug fix release 1.21.2?

The way the issue is currently phrased makes it sound like a feature request (which means that it needs to go in with a minor release). If we retitled it to something like "Restore `original` property for ...", then it would make it a task (eligible for a bug fix release), and if we retitled it to something like "`original` property missing from ...", then it would be a bug (again eligible for a bug fix release) 😉

**Comment by argiepiano** [comment]
> So a @var Node for the $node->original would make it so that node-only properties and methods would auto complete. 

Got it. I've added the property to all 5 core entity definitions. 

> Seems this might be considered a bug. Should it go into the next bug fix release 1.21.2?

I'm OK with this going into a bug fix release or a minor release. We solved the original need for this addition  (for Rules) in a different way, but it'll be good to have this property again available for TaxonomyVocabulary.

**Comment by argiepiano** [comment]
@indigoxela you are right. @quicksketch this is the result of adding that property to all entity classes. It may be harmless, but perhaps to be on the safe side we can unset that property from the original object when adding it to the changed one.

**Comment by argiepiano** [comment]
I've added a commit to the PR to unset the `original` property in the original entity and vocabulary.

**Comment by argiepiano** [comment]
@quicksketch I'm having strong doubts about your suggestion of adding $original as a property of entity classes and the vocabulary class. Now I really think it's not a good idea, and it will break many places.

Most contrib (and perhaps even core) modules that use the hook_XXX_presave() check if $original is set. (Notice that presave hooks are invoke both when a new entity is being saved, and when an existing entity is being updated). Now that this is a public property  of the class itself, this means that this check will always give true (it is always set now). A solution may be for those places to check for !empty() - BUT this is just too risky a change for too little a gain (just to get the IDE to recognize $original).

Take for example [this line](https://github.com/backdrop-contrib/rules/blob/dd4cfdb832624731da5bc4940862202a7a631910/modules/events.inc#L26) in Rules. There are other examples like this one. This line will now always be true, since $original is always set, even when the entity is brand new and $original is empty. 

I'll wait for your input before reverting that addition.

**Comment by quicksketch** [comment]
Thanks @argiepiano for your patience. Wouldn't be the first time I derailed something. :zany_face: 

I don't think the example you mentioned would be affected. Both `isset()` and `empty()` require a non-`NULL` value. Simply having the property exist would still return `FALSE` for `isset()` checks. But if some code were to use `property_exists()` then they would indeed get a different result.

Take this example code for a full example:

```php
$object = new stdClass();
$object->original = NULL;
if (isset($object->original)) {
    print 'isset'; // Never is called.
};
if (property_exists($object, 'original')) {
    print 'property exists'; // Always is called.
}
```

You can verify the output of this on all PHP versions at https://3v4l.org/jofWf

You might still be right that excluding this property everywhere might be better anyway. Perhaps we should instead not define it, and introduce a `getOriginal()` method instead (in a separate issue).

**Comment by argiepiano** [comment]
Oops, you are right about isset  😅  Should have known that! Thanks for taking the time to explain and create an example.

So, I'm not sure then what's the best option, whether to define it or not. I'll let you and others chime in.

**Comment by indigoxela** [comment]
TBH, my personal preference would be to reduce this PR here to cover the initial task - provide the original vocab only.

Adding the property always in all entities (and also vocabs) might be better discussed in a follow-up, as it turned out that there *could* be negative side effects. In that follow-up we can also discuss, if we should introduce a new method getOriginal().

**Comment by klonos** [comment]
What @indigoxela said ^^ seems the best way forward to me 👍🏼

**Comment by argiepiano** [comment]
OK, thank you @indigoxela , @klonos and @quicksketch. I have reverted the last changes, and only left there the first commit that added $original to TaxonomyVocabulary (without declaring it as a property). Let me know what you think.

**Comment by indigoxela** [comment]
> Let me know what you think.

I think, that's the safer choice. And of course, it works for me.

Should we create a follow-up for the default property and/or new method?

**Comment by quicksketch** [comment]
Thanks @argiepiano and @indigoxela. I think you're both right that this got out of hand. Going back to the simplified approach and separating out the methods/properties into a follow-up is a much better idea to get this unblocked.

In the end, I merged https://github.com/backdrop/backdrop/pull/3920 into both 1.x and 1.21.x. I think it can qualify as a bug that Taxonomies act differently than every other core entity and adversely affect token availability.

**Comment by klonos** [comment]
> Should we create a follow-up for the default property and/or new method (also for entities)?

> ... separating out the methods/properties into a follow-up is a much better idea ...

@argiepiano or @indigoxela do you feel like creating that follow-up ticket? You'd be much more qualified to put together a proper issue summary that I.

**Comment by indigoxela** [comment]
> do you feel like creating that follow-up ticket?

Sure, here it is:  #5500
